lib/elixir/lib/module/types/infer.ex

```diff
@@ -467,6 +467,8 @@ defmodule Module.Types.Infer do
   def unify_kinds(:optional, :optional), do: :optional
 
   # Collect relevant information from context and traces to report error
+  # TODO: We should do this lazily since in some cases unification will error
+  #       but we continue attempting unifying other types
```

Comment:
During map unification we may make many attempts to unify pairs. During the failed attempts it is unnecessary to collect information from the context since the error will not be used.